File: 0918-maximum-sum-circular-subarray/0918-maximum-sum-circular-subarray.kt
class Solution {
    fun maxSubarraySumCircular(nums: IntArray): Int {
        var array_sum:Int = 0
        var (currMax, currMin) = Int.MIN_VALUE to Int.MAX_VALUE
        var (maxSoFar, minSoFar) = 0 to 0
        
        val arraySize = nums.size

        for(i in nums) {
            array_sum = array_sum + i
            
            maxSoFar = maxSoFar + i
            currMax = maxOf(maxSoFar, currMax)
            
            minSoFar = minSoFar + i
            currMin = minOf(minSoFar, currMin)
            if (maxSoFar < 0)
                maxSoFar = 0
            if (minSoFar > 0)
                minSoFar = 0
        }
        
        return if (currMin == array_sum) currMax else maxOf(currMax, array_sum - currMin)
    }
}

File: 0918-maximum-sum-circular-subarray/0918-maximum-sum-circular-subarray.py
import math

class Solution:
    def maxSubarraySumCircular(self, nums: List[int]) -> int:
        array_sum = 0
        curr_max = -math.inf
        max_so_far = -math.inf
        curr_min = math.inf
        min_so_far = math.inf

        # Concept of Kadane's Algorithm
        for i in range(len(nums)):

            # Compute sum of complete array
            array_sum += nums[i]
            # print("array sum", array_sum)

            # Kadane's Algorithm to find Maximum subarray sum.
            curr_max = max(curr_max + nums[i], nums[i])
            # print("curr_max", curr_max)
            max_so_far = max(max_so_far, curr_max)
            # print("max_so_far", max_so_far)

            # Kadane's Algorithm to find Minimum subarray sum.
            curr_min = min(curr_min + nums[i], nums[i])
            # print("curr_min", curr_min)
            min_so_far = min(min_so_far, curr_min)
            # print("min_so_far", min_so_far)
            
        if (array_sum == min_so_far):
            return max_so_far
        
        return max(max_so_far, array_sum - min_so_far)